src/mars.c

```diff
@@ -1694,21 +1710,74 @@ Language changes listed by -transition=id:\n\
                 fprintf(global.stdmsg, "code      %s\n", m->toChars());
 
             obj_start(m->srcfile->toChars());
-            genObjFile(m, global.params.multiobj);
+            genObjFile(m, false);
             if (entrypoint && m == rootHasMain)
-                genObjFile(entrypoint, global.params.multiobj);
+                genObjFile(entrypoint, false);
             for (size_t j = 0; j < Module::amodules.dim; j++)
             {
+                // todo: This part is not the best for the total size of object files.
+                // For example:
+                // When both of main.d and code.d import mx, even if main.obj
+                // already contains mx->marray, it is stored in code.obj again.
                 Module *mx = Module::amodules[j];
-                if (mx != m && mx->importedFrom == m && (mx->marray || mx->massert || mx->munittest))
-                    genhelpers(mx, true);
```

Comment:
Ah, so #3552 is where it came from. I did not review that PR. I believe it is the wrong solution, however. A very simple solution is to compile the helpers and moduleinfo into stdio.obj. Then, you have to link with stdio.obj. I do not think it should be a surprise to anyone that if they import stdio, they have to link with stdio.obj. (The stdio.obj should be in the library, too, meaning it will get pulled in automatically.)

This solution resolves https://issues.dlang.org/show_bug.cgi?id=846 in a simple, easy to understand, and easy to implement manner. Furthermore, it eliminates all that .obj file bloat I've noticed where the helper functions for every import get generated into every .obj file.